### DISCUSSION TITLE
Summary / aggregation phase options

### DISCUSSION THREAD

**Discussion by ghannington** [genevaers/community#260] [Ideas]
I am working on the short term proposal to simplify the old Format Phase routines by removing Reports.
While working on this I see in the assembler code that there appears to be options for column calculations to be executed before summarisation, after summarisation , and both before and after. 

This does not make much sense now with the proposal that Reports, and multi-level aggregation, are removed.

With the new simplified summary routine I propose we only support calculations before summarisation (detailed level), or after summarisation (key break level). The Workbench should have an indicator to state when the calculation is to be performed.

The following record aggregation options are also proposed (from discussions with Neil and Randall): SUM, MIN, MAX, COUNT, AVERAGE.

In the Workbench the Group Aggregation row will not be required. Also the Report option on the General tab, and the associated Report tab will not be required.

**Comment by ghannington** [comment]
Something I have discovered during testing;
The option 'Zero-Value Record Suppression' is on the 'Format Phase' tab in the Workbench, but testing this option, it seems to apply to Reports only. If this is the case, then it should be removed from the Workbench too. 
Any thoughts? Should this option apply to detail and summary outputs (non-Report)?

**Comment by neilbeesley** [comment]
Some other aggregation functions that could be added later are:
MEDIAN
STD DEV
CORRELATION
COUNT DISTINCT
VARIANCE

**Comment by ghannington** [comment]
COUNT and MEAN (AVERAGE)

There are issues when the View has Extract-Phase aggregation switched on. Currently there are no counts kept in the extract record to indicate how many records have been aggregated in each extract record. 
To calculate the MEAN at summary time we need to know the record count for each sort key break.
There is no spare space in the Extract record header, so we would either have to change the layout of the Extract record, or have the count in a CT column.
For this second case we could:
- enforce in the Workbench (and MR91) that if a MEAN column is specified, there must be a COUNT column, or
- if there is a MEAN column defined, then MR91 could add in a COUNT column under the covers as a non-displayable column.

Any thoughts or comments on this?

**Comment by rhness** [comment]
For reference, here are the aggregate functions that Db2 offers: 

https://www.ibm.com/docs/en/db2-for-zos/13?topic=functions-aggregate
